### PR TITLE
Fix primaries on FirefoxOS

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@jedp - this is what I am thinking so far.

### CODE REVIEW

File: resources/static/authentication_api.js [hunk 1]
```diff
@@ -33,16 +33,25 @@
       setTimeout(function() { cb(email); }, 0);
     };
 
+    var ipServer = 'https://login.persona.org';
+    if (navigator.userAgent ===
+          "Mozilla/5.0 (Mobile; rv:18.0) Gecko/18.0 Firefox/18.0") {
```
Comment (jedp) [inline resources/static/authentication_api.js]: I guess we can just check that it matches "Mobile" and "Gecko"?

**Comment by jedp** [comment]
That is pretty damn cool

File: resources/static/authentication_api.js (hunk 1 above)
Comment (seanmonstar) [inline resources/static/authentication_api.js]: Zoh my gawd, why? We can't feature detect FirefoxOS?

Comment (shane-tomlinson) [inline resources/static/authentication_api.js]: @seanmonstar - proof of concept, don't sweat yet! Do you have a feature detection snippet for FFOS?

Comment (shane-tomlinson) [inline resources/static/authentication_api.js]: I suppose we could check for navigator.mozId

Comment (jedp) [inline resources/static/authentication_api.js]: maybe check for navigator.mozId and scan for Mobile in the UA string.  I don't know of a way to feature detect for fxos specifically

File: lib/configuration.js [hunk 2]
```diff
@@ -393,6 +393,7 @@ module.exports.performSubstitution = function(app) {
     app.use(postprocess(function(req, buffer) {
       return buffer.toString()
         .replace(new RegExp('https://login.persona.org', 'g'), conf.get('public_url'))
+        .replace(new RegExp('https://firefoxos.persona.org', 'g'), conf.get('public_url'))
```
Comment (shane-tomlinson) [inline lib/configuration.js]: the replacement should be pointing to the cname instead of public_url.

**Comment by ozten** [comment]
Heroic!

File: lib/configuration.js [hunk 3]
```diff
@@ -393,6 +398,7 @@ module.exports.performSubstitution = function(app) {
     app.use(postprocess(function(req, buffer) {
       return buffer.toString()
         .replace(new RegExp('https://login.persona.org', 'g'), conf.get('public_url'))
+        .replace(new RegExp('https://firefoxos.persona.org', 'g'), conf.get('firefoxos_url'))
```
Comment (jedp) [inline lib/configuration.js:406]: How about something like `https://firefoxos\.(?:persona|(?:dev\.)?anosrep)\.org`, so we work with dev and stage, too?

Comment (shane-tomlinson) [inline lib/configuration.js:406]: We always keep firefoxos.persona.org within the code, postprocess takes care of updating the firefoxos_url based on the environment.

Comment (jedp) [inline lib/configuration.js:406]: Ah ok.  That makes sense.  Thanks for the comments in the code about this.

File: lib/configuration.js [hunk 4]
```diff
@@ -360,14 +365,18 @@ if (conf.get('env') === 'test_json') {
...
 
-// if static and verifier urls are not explicitly set, default them to the
-// same as the public url (used in ephemeral and local deployments)
+// if static, verifier, and firefoxos urls are not explicitly set,
+// default them to the same as the public url (used in ephemeral
+// and local deployments)
 if (!conf.has('public_static_url') || conf.get('public_static_url') === null) {
   conf.set('public_static_url', conf.get('public_url'));
 }
 if (!conf.has('public_verifier_url')) {
```
Comment (jedp) [inline lib/configuration.js:374]: Does this line also want the `|| conf.get('public_verifier_url') === null` test, like the others?

File: resources/static/authentication_api.js [hunk 5]
```diff
@@ -33,16 +33,28 @@
       setTimeout(function() { cb(email); }, 0);
     };
 
+    var ipServer = 'https://login.persona.org';
+    // FirefoxOS UA string:
+    // "Mozilla/5.0 (Mobile; rv:18.0) Gecko/18.0 Firefox/18.0
+    // Android UA string:
+    // Mozilla/5.0 (Android; Mobile; rv:18.0) Gecko/18.0 Firefox/18.0
+    if (!!navigator.mozId && (navigator.userAgent.indexOf("(Mobile;") > -1)) {
```
Comment (jedp) [inline resources/static/authentication_api.js:41]: Maybe we can query a property on `navigator.mozSettings`.  It's what the Settings app looks at to get the OS version etc.  I'm trying to find out whether there's a stable way to do that, with which vendors cannot interfere.

Comment (jedp) [inline resources/static/authentication_api.js:41]: Oho! Or maybe UA sniffing is precisely what we want to do.  [Partners are not allowed to modify the FirefoxOS UA string](https://wiki.mozilla.org/B2G/User_Agent/Partner_Changes_Policy).

File: lib/configuration.js (hunk 4 above)
Comment (shane-tomlinson) [inline lib/configuration.js:374]: I don't believe so, public_verifier_url is not defined in production.js and does not need to be overridden. I left this code as is.

File: resources/static/authentication_api.js (hunk 5 above)
Comment (shane-tomlinson) [inline resources/static/authentication_api.js:41]: @jedp - do you know what we can get out of navigator.mozSettings?

Comment (jedp) [inline resources/static/authentication_api.js:41]: All the device info you can see in the Settings app is taken from `navigator.mozSettings`.  I was surprised, though, when talking with someone on Deutsche Telekom's QA team to find that they don't have quite the same set of things we have our our dev builds.  I can look into this further to find out what the real story is.

**Comment by jedp** [comment]
So, this is awesome.

**Pull request merged**